app/src/main/java/com/github/ravifrancesco/softwaredevproject/Point.java

```diff
@@ -139,4 +144,171 @@ public double getAltitude() {
+        // Calculate norm(b-a), where norm divides a vector by its length to produce a unit vector.
+        double dx = b.getLatitude() - a.getLatitude();
+        double dy = b.getLongitude() - a.getLongitude();
+        double dz = b.getAltitude() - a.getAltitude();
+        double dist2 = dx*dx + dy*dy + dz*dz;
+        double dist = Math.sqrt(dist2);
+        return new Point(dx/dist, dy/dist, dz/dist);
+    }
+
+    /**
+     * Compute the location to the point
+     * @param c Point that gets computed
+     * @return new Point with computed location
+     */
+    private Pair<Point, Point> locationToPoint(Point c){
+        // Convert (lat, lon, elv) to (x, y, z).
+        double lat = c.getLatitude() * Math.PI / 180.0;
+        double lon = c.getLongitude() * Math.PI / 180.0;
+        double elv = c.getAltitude();
+
+        double radius = earthRadiusInMeters(lat);
+        double clat   = geocentricLatitude(lat);
+
+        double cosLon = Math.cos(lon);
+        double sinLon = Math.sin(lon);
+
+        double cosLat = Math.cos(clat);
+        double sinLat = Math.sin(clat);
+        double x = radius * cosLon * cosLat;
+        double y = radius * sinLon * cosLat;
+        double z = radius * sinLat;
+
+        // We used geocentric latitude to calculate (x,y,z) on the Earth's ellipsoid.
+        // Now we use geodetic latitude to calculate normal vector from the surface, to correct for elevation.
+        double cosGlat = Math.cos(lat);
+        double sinGlat = Math.sin(lat);
+
+        double nx = cosGlat * cosLon;
+        double ny = cosGlat * sinLon;
+        double nz = sinGlat;
+
+        x += elv * nx;
+        y += elv * ny;
+        z += elv * nz;
+
+        return new Pair<>(new Point(x, y, z), new Point(nx, ny, nz));
+    }
+
+    /**
+     * Calculates the earth radius given the latitude
+     * See http://en.wikipedia.org/wiki/Earth_radius
+     * @param latitudeRadians latitude in radians (geodetic)
+     * @return Radius of the earth in meters
+     */
+    private double earthRadiusInMeters(double latitudeRadians){
+        double a = 6378137.0;  // equatorial radius in meters
+        double b = 6356752.3;  // polar radius in meters
+        double cos = Math.cos (latitudeRadians);
+        double sin = Math.sin (latitudeRadians);
+        double t1 = a * a * cos;
+        double t2 = b * b * sin;
+        double t3 = a * cos;
+        double t4 = b * sin;
+        return Math.sqrt ((t1*t1 + t2*t2) / (t3*t3 + t4*t4));
+    }
+
+    /**
+     * Converts a geodetic latitude (like GPS values) to a geocentric latitude (angle measured
+     * from center of Earth between a point and the equator).
+     * See https://en.wikipedia.org/wiki/Latitude#Geocentric_latitude
+     * @param lat Geodetic latitude in radian
+     * @return Converted latitude in radian to geocentric latitude
+     */
+    private double geocentricLatitude(double lat){
+        double e2 = 0.00669437999014;
```

Comment:
Magic numbers, should this be a constant?

Reply:
Changed to constants